### ISSUE TITLE
Do floating point addition and subtraction in decimal

### ISSUE THREAD

**Issue by zathras** [zathras/jrpn#78]
Thinking about #76 and #74 some more, I think it's worth doing addition and subtraction in decimal floating point, as a special case.  This is in the same category as e.g. cos(90 degrees), in that it can be highly visible and counter-intuitive.

Take, for example. 1.0000001 - 1.  In decimal floating point (like the real 15C does), that yields 1e-7.  In binary floating point, it yields a number that rounds to 1.000000001E-07 in ten-digit mantissa decimal floating point.  This is annoying, and most people won't expect it.

**Comment by zathras** [comment]
On a real 15C, it looks like it should be adequate to carry out the operations on a 12 digit mantissa (truncating digits that fall off the right hand side), and then rounding the result away from zero.  Here are some examples carried out on the real calculator to discover this (and, later, test the simulator):

1 + 5e-10 yields 1.000000001
1 - 5e-10 yields 0.9999999995
1 - 6e-11 yields 0.9999999999
1 - 5e-11 yields 1
1 - 5.000000001e-11 yields 0.9999999999
2 + 5e-10 yields 2.000000001
2 - 5e-10 yields 2
3 + 5e-10 yields 3.000000001
3 - 5e-10 yields 3

-1 - 5e-10 yields -1.000000001
-1 + 5e-10 yields -0.9999999995
-1 + 6e-11 yields -0.9999999999
-1 + 5e-11 yields -1
-1 + 5.000000001e-11 yields -0.9999999999
-2 - 5e-10 yields -2.000000001
-2 + 5e-10 yields 2
-3 - 5e-10 yields -3.000000001
-3 + 5e-10 yields -3

**Comment by pane** [comment]
Will have to try it on my real calculator (when back home). I just looked at the [HP Solve Issue #26](http://h20331.www2.hp.com/hpsub/downloads/HP_Calculator_eNL_01_January_2012.pdf) about guard digits, rounding and reusing results in HP calculators. When I tried first your JRPN I was surprised how you achieved exact legacy forensic result 9 sin cos tan atan acos asin  getting 9.000417403 and 100-99,99-0,01 not yielding 5,11569953E-15 (usual signs of binary calculation usage).
Looking for your future enhancement.
KR
P

**Comment by zathras** [comment]
> When I tried first your JRPN I was surprised how you achieved exact legacy forensic result 9 sin cos tan atan acos asin getting 9.000417403 and 100-99,99-0,01 not yielding 5,11569953E-15 (usual signs of binary calculation usage).

Interesting.  What's going on here is that I'm not keeping IEEE doubles in the calculator memory.  When I put a result on the stack, it's converted to a decimal floating point with 10 digits of mantissa.  So, 100 - 99.99 gives (about) 0.010000000000005116, which then gets rounded to ten digits, which is exactly 0.01.  Then in the second subtraction, 0.01 is converted to a double that's about 0.01, and the second argument is converted to a double that's about 0.01 _and is exactly the same as the first double_, resulting in zero.

In other words, the secret is rounding the result to a ten-digit decimal mantissa after every operation.

That doesn't break down until you get more digits, like 100 - 99.999999.  That's 1e-6, but carried out in IEEE double binary floating point and then rounded to ten decimal digits, it comes out to  9.999999975E-07.  Close, but no cigar!

Besides subtraction (and addition of numbers with differing signs), I haven't been able to think of any operations where doing the internal calculation in 64 bit binary would present this kind of issue at the margins.  Let me know if you think of any!

Every time I do discrete math, it seems like there's at least one surprise.  Early in my career in the '90s I did a C++ decimal number package.  I implemented division, thinking "how hard can it be?"  The answer turned out to be "pretty hard."

**Comment by pane** [comment]
It make sense! When converting 0,01 from the result and than "different" 0,01 from entry make the same FP number, wow.
It looks to me that the closest approximation to original behavior would be to truncate (or round?) to 13 significant digits, calculate with FP precision, and than round to 10 digits and discard any leftover. I am still wondering how you managed that 9 sin cos tan ... example to be correct (will try some rounding in Excel since it gives 5,11569953E-15 for 100-99,99-0,01).

I recollect some weird behavior when using ->H.MS conversion with wrong rounding (like showing 10H 10MIN **60**SEC) and will try to look for it.

Also something to [division](https://youtu.be/7Kd3R_RlXgc?si=L9drWSpE6hQI-4dj).

**Comment by zathras** [comment]
Thanks.  I think I'm on pretty solid ground with ->H.MS.  One fun thing about that on the real calculator (and the simulator) is that it's easy to get it to _display_ 1.5960.  1.999999999 -> H.MS will do that, because the answer is 1 hour, 59 minutes, 59.99999 seconds, which rounds to 1.5960 in FIX 4.

I don't know if you're building from source, but if you are, I just put back a first cut at decimal addition/subtraction.

Cute video about division!

**Comment by zathras** [comment]
https://github.com/zathras/jrpn/commit/126e91143fa5200264ded98e00c5dd674e039a45

**Comment by pane** [comment]
I am using yours builds. It looks like a lot of effort going in! Hope you enjoy this challenge.

I am just rereading hpforum [H.MS thread](https://www.hpmuseum.org/forum/thread-11352.html) and tried this:
0.08 ->H =  0.133333333
->H.MS =  0.075999990

HP15C CE gives back 0.080000000. There must be some special addition in H.MS treatment.

**Comment by zathras** [comment]
> I am using yours builds. It looks like a lot of effort going in! Hope you enjoy this challenge.

Discrete math is always fun!  Do you mean you're using the build from the github page?  Which platform(s)?

> HP15C CE gives back 0.080000000. There must be some special addition in H.MS treatment.

Interesting!  

I did a quick check, and 9e98 ->H.MS doesn't overflow, which is another data point.

Hmmm..

>>> 0.1333333333 * 60
7.9999999980

That rounds to exactly 8 (minutes).  This is of course a less precise answer than 7 minutes 59.9999 seconds, but it's the one you get with a more primitive algorithm with intermediate results rounded to ten digits.  I bet that's what the real 15C is doing.

Well, that's actually easier to implement than what I was doing!

**Comment by zathras** [comment]
Oh, not quite right.  This yields the more precise result:
```
  final double sign = m.x.isNegative ? -1 : 1;
  final Value oneTimesSign =
      m.x.isNegative ? Value.oneF.negateAsFloat() : Value.oneF;
  Value hr = m.x.intOp();
  final Value rest = Value.fromDouble(60 * m.x.fracOp().asDouble);
  Value min = rest.intOp();
  Value sec = Value.fromDouble(rest.fracOp().asDouble * 60);
  //
  // Now, check for min or sec >= 60.  I suspect this may be impossible, but
  // it's easier to write this code than it is to convince myself with 100%
  // certainty :-)
  //
  if (sec.asDouble.abs() >= 60) {
    sec = sec.decimalSubtract(Value.fromDouble(sign * 60.0));
    min = min.decimalAdd(oneTimesSign);
  }
  if (min.asDouble.abs() >= 60) {
    min = min.decimalSubtract(Value.fromDouble(sign * 60.0));
    hr = hr.decimalAdd(oneTimesSign);
  }
  m.resultX = hr.decimalAdd(min.timesTenTo(-2)).decimalAdd(sec.timesTenTo(-4));
```

To get the same result as the real 15C, I need to first multiply by 60 (or by 3600), and then do mod operations on the remaining integer part.  Huh.

**Comment by zathras** [comment]
Hang on, I miscounted.  7.9999999980 in ten digits is 7.999999998.  So that doesn't explain it.

I also note that 1.999999999 converting to 1 hour, 59 minutes, 59.99999 (which the real 15C does) is a worse result than converting to 2 hours.  
59.999996400 is the real number of seconds, and
59.99999 is what the 15C gives.

OK, so the 15C is doing a worse job of ->H.MS than I am, but my existing code manages to be bug-compatible for the case of 1.999999999.  I'm not quite sure how I feel about that :-)

**Comment by pane** [comment]
Do you mean you're using the build from the github page? Which platform(s)?

**Comment by pane** [comment]
I am using Windows app , browser through web https://jrpn.jovial.com/ and Android (some shows version 2.1.10, some 2.1.11)

Now I see how you got to correct 9 sin cos tan ... (and found some inaccuracy in XLS of arcos while checking 12 significant digit result and how Excel ROUND works).

KR
P

**Comment by pane** [comment]
> I am using Windows app , browser through web https://jrpn.jovial.com/ and Android (some shows version 2.1.10, some 2.1.11)
> 

I see you have done 2.1.12 beta1. I was not able to update Android apps with apk files downloaded to the phone (I can get over the "warning being the app from other source" still can chose install anyway so I have correctly allowed installations from other sources but than it says it is blocked (Pixel 4a Android 13).
Similar for Windows with the message that certificate is not valid.

Is there an easy way to produce the link to start a beta version in the browser?

Looking forward to test new version.
P

**Comment by zathras** [comment]
I'm a little hesitant to make the web version be a beta, but if necessary I will.  But could you try the following first?

I see that a .msix file is just a zip file, with some extra stuff in it.  So that means you can download a .msix, rename it something.zip, and unzip it.  So, maybe try unzipping the msix somewhere, and running jrpn15c.exe?  You might need to open a console window, make the current working directory be the one containing the .exe, and launch it from the command line.

If that doesn't work for you, let me know and I'll just make the web version be the beta.

**Comment by pane** [comment]
> I see that a .msix file is just a zip file, with some extra stuff in it. So that means you can download a .msix, rename it something.zip, and unzip it. So, maybe try unzipping the msix somewhere, and running jrpn15c.exe? 

I did not think of it, working nicely, thank you for the hint. Capital C for HP16C is an easy version identifier!

What timezone you are in (CET for me).

KR
P

**Comment by zathras** [comment]
> What timezone you are in (CET for me).

US West Coast for me.  Just moved back from The Philippines, where I was a US Peace Corps volunteer.

**Comment by pane** [comment]
> > 
> 
> Besides subtraction (and addition of numbers with differing signs), I haven't been able to think of any operations where doing the internal calculation in 64 bit binary would present this kind of issue at the margins. Let me know if you think of any!

I like your improvements in decimal calculations, works well!

Another small observation. I think this is not related to binary or decimal calculation but your results differ from real HP15C - although yours's are more precise, no complains:
In RAD mode: 
SIN (1e22) (more in this [thread](https://hpmuseum.org/forum/thread-20577.html))
SIN (3.141592653) (as in this [thread](https://hpmuseum.org/forum/thread-2971.html))

Very nice and useful toys, indeed!

KR
P

**Comment by zathras** [comment]
Published in 2.1.12
